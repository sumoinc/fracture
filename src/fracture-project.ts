import {
  NodeProject,
  NodeProjectOptions,
  Prettier,
} from "projen/lib/javascript";
import { SetOptional, SetRequired } from "type-fest";
import { Environment } from "./environments";

export type FractureProjectOptions = SetRequired<
  SetOptional<NodeProjectOptions, "defaultReleaseBranch">,
  "outdir"
>;

/**
 * Used by subclasses to specify deploy location(s) for each project
 */
export interface DeployOptions {
  /**
   * The branch prefix this deployment is targeting.
   *
   * @default - defaultReleaseBranch found in Settings()
   */
  readonly branchPrefix?: string;

  /**
   * Environemnt to deploy into.
   */
  readonly environment: Environment;
}

export class FractureProject extends NodeProject {
  constructor(
    public readonly parent: NodeProject,
    options: FractureProjectOptions
  ) {
    super({
      parent,
      defaultReleaseBranch: "main",
      // inherit from parent project
      license: parent.package.license,
      prettier: parent.prettier && parent.prettier instanceof Prettier,
      packageManager: parent.package.packageManager,
      pnpmVersion: parent.package.pnpmVersion,
      ...options,
    });

    // don't package
    this.packageTask.reset();
    // don't allow default to run in subprojects, otherwise it runs root and
    // causes unwanted recusion.
    this.defaultTask?.reset();
  }
}
