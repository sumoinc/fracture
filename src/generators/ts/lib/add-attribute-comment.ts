import { Attribute } from "../../../model";
import { TypeScriptSource } from "../typescript-source";

export const addAttributeComment = (f: TypeScriptSource, a: Attribute) => {
  f.line(`/**`);
  a.comment.forEach((c) => f.line(` * ${c}`));
  f.line(` */`);
};
