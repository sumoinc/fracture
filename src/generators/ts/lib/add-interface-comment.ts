import { Shape } from "../../../model";
import { TypeScriptSource } from "../typescript-source";

export const addInterfaceComment = (f: TypeScriptSource, e: Shape) => {
  f.line(`/**`);
  e.comment.forEach((c) => f.line(` * ${c}`));
  f.line(` */`);
};
