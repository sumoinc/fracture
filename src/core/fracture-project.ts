import { NodePackageManager } from "projen/lib/javascript";
import {
  TypeScriptProject,
  TypeScriptProjectOptions,
} from "projen/lib/typescript";
import { deepMerge } from "projen/lib/util";
import { SetOptional } from "type-fest";
import { FracturePackage } from "./fracture-package";
import { PnpmWorkspace } from "../pnpm/pnpm-workspace";
import { TurboRepo } from "../turborepo/turbo-repo";

/**
 *
 * Top level options for a Fracture project.
 *
 */
export interface FractureProjectOptions extends TypeScriptProjectOptions {
  packageDir?: string;
  appDir?: string;
}

/**
 * The root of the entire application.
 */
export class FractureProject extends TypeScriptProject {
  // member components
  public fracturePackages: FracturePackage[] = [];
  private turborepo: TurboRepo;
  // all other options
  public readonly options: Required<FractureProjectOptions>;

  constructor(
    options: SetOptional<FractureProjectOptions, "defaultReleaseBranch">
  ) {
    const defaultOptions: Partial<FractureProjectOptions> = {
      defaultReleaseBranch: "main",
      packageDir: "packages",
      appDir: "apps",
    };

    const requiredOptions: Partial<FractureProjectOptions> = {
      packageManager: NodePackageManager.PNPM,
      pnpmVersion: "8",
      prettier: true,
      projenrcTs: true,
      deps: ["turbo"].concat(options.deps ?? []),
    };

    const mergedOptions = deepMerge([
      defaultOptions,
      options,
      requiredOptions,
    ]) as Required<FractureProjectOptions>;

    super(mergedOptions);
    this.options = mergedOptions;

    // configure workspace
    new PnpmWorkspace(this);

    // configure turborepo
    this.turborepo = new TurboRepo(this);
  }

  public get packageDir(): string {
    return this.options.packageDir;
  }

  public get appDir(): string {
    return this.options.appDir;
  }

  public get buildTask() {
    return TurboRepo.buildTask(this);
  }

  /**
   * Build the project.
   *
   * Call this when you've configured everything, prior to preSynthesize.
   *
   * @returns void
   */
  public build() {
    this.fracturePackages.forEach((p) => {
      p.build();
    });
    this.turborepo.build();
  }
}
