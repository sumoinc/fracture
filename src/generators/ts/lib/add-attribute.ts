import { addAttributeComment } from "./add-attribute-comment";
import { Attribute } from "../../../model";
import { TypeScriptSource } from "../typescript-source";

export const addAttribute = (f: TypeScriptSource, a: Attribute) => {
  addAttributeComment(f, a);

  const q = a.isRequired ? "" : "?";
  f.line(`${a.name}${q}: ${a.typeScriptType};`);
};
