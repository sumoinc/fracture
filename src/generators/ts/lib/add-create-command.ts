import { formatStringByNamingStrategy } from "../../../core/naming-strategy";
import { Shape } from "../../../model";
import { TypeScriptSource } from "../typescript-source";

export const addCreateCommand = (e: Shape) => {
  const command = `${e.fracture.namingStrategy.operations.commands.commandPrefix}-${e.fracture.namingStrategy.operations.crud.createName}-${e.name}-${e.fracture.namingStrategy.operations.commands.commandSuffix}`;
  const commandInput = `${e.fracture.namingStrategy.operations.commands.inputPrefix}-${e.fracture.namingStrategy.operations.crud.createName}-${e.name}-${e.fracture.namingStrategy.operations.commands.inputSuffix}`;
  const commandOutput = `${e.fracture.namingStrategy.operations.commands.outputPrefix}-${e.fracture.namingStrategy.operations.crud.createName}-${e.name}-${e.fracture.namingStrategy.operations.commands.inputSuffix}`;

  const fileName = formatStringByNamingStrategy(
    command,
    e.fracture.namingStrategy.ts.file
  );
  const f = new TypeScriptSource(e.service, `/commands/${fileName}.ts`);

  const commandInputName = formatStringByNamingStrategy(
    commandInput,
    e.fracture.namingStrategy.model.shapeName
  );

  const commandOutputName = formatStringByNamingStrategy(
    commandOutput,
    e.fracture.namingStrategy.model.shapeName
  );

  f.open(`export interface ${commandInputName} {`);
  f.line("foo: string;");
  f.close(`}`);
  f.line("\n");

  f.open(`export interface ${commandOutputName} {`);
  f.line("foo: string;");
  f.close(`}`);
  f.line("\n");

  f.open(`export const handler = async (event: any) => {`);
  f.line("return true;");
  f.close(`};`);
  f.line("\n");
};
