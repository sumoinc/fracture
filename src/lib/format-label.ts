import { camelCase, pascalCase } from "change-case";
import { ValueOf } from "type-fest";
import { NamingStrategyType } from "../core";

export const formatLabel = (
  s: string,
  ns: ValueOf<typeof NamingStrategyType>
) => {
  switch (ns) {
    case NamingStrategyType.PASCAL_CASE:
      return pascalCase(s);
      break;
    case NamingStrategyType.CAMEL_CASE:
      return camelCase(s);
    default:
      throw new Error(`Invalid naming strategy ${ns}`);
      break;
  }
};
