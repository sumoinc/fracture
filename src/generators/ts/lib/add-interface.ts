import { addAttribute } from "./add-attribute";
import { addInterfaceComment } from "./add-interface-comment";
import { Entity } from "../../../model";
import { TypeScriptSource } from "../typescript-source";

export const addInterface = (f: TypeScriptSource, e: Entity) => {
  addInterfaceComment(f, e);

  f.open(`export interface ${e.name} {`);
  e.attributes.forEach((a) => {
    addAttribute(f, a);
  });
  f.close(`}`);
  f.line("\n");
};
