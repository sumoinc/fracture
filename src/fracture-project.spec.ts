import { NodeProject } from "projen/lib/javascript";
import { FractureProject } from "./fracture-project";

describe("success conditions", () => {
  test("Smoke test", () => {
    const root = new NodeProject({
      name: "my-project",
      defaultReleaseBranch: "main",
    });
    const fractureProject = new FractureProject(root, {
      name: "my-sub-project",
      outdir: "sites/foo",
    });
    expect(fractureProject).toBeTruthy();
  });
});
