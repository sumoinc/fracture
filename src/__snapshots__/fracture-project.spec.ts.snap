// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`package.json 1`] = `
Object {
  "//": "~~ Generated by projen. To modify, edit .projenrc.ts and run \\"npx projen\\".",
  "dependencies": Object {
    "@sumoc/fracture": "*",
  },
  "devDependencies": Object {
    "@types/jest": "*",
    "@types/node": "^16",
    "@typescript-eslint/eslint-plugin": "^6",
    "@typescript-eslint/parser": "^6",
    "constructs": "^10.0.0",
    "eslint": "^8",
    "eslint-config-prettier": "*",
    "eslint-import-resolver-node": "*",
    "eslint-import-resolver-typescript": "*",
    "eslint-plugin-import": "*",
    "eslint-plugin-prettier": "*",
    "jest": "*",
    "jest-junit": "^15",
    "npm-check-updates": "^16",
    "prettier": "*",
    "projen": "*",
    "standard-version": "^9",
    "ts-jest": "*",
    "ts-node": "*",
    "typescript": "*",
  },
  "jest": Object {
    "clearMocks": true,
    "collectCoverage": true,
    "coverageDirectory": "coverage",
    "coveragePathIgnorePatterns": Array [
      "/node_modules/",
    ],
    "coverageReporters": Array [
      "json",
      "lcov",
      "clover",
      "cobertura",
      "text",
    ],
    "globals": Object {
      "ts-jest": Object {
        "tsconfig": "tsconfig.dev.json",
      },
    },
    "preset": "ts-jest",
    "reporters": Array [
      "default",
      Array [
        "jest-junit",
        Object {
          "outputDirectory": "test-reports",
        },
      ],
    ],
    "testMatch": Array [
      "<rootDir>/src/**/__tests__/**/*.ts?(x)",
      "<rootDir>/(test|src)/**/*(*.)@(spec|test).ts?(x)",
    ],
    "testPathIgnorePatterns": Array [
      "/node_modules/",
    ],
    "watchPathIgnorePatterns": Array [
      "/node_modules/",
    ],
  },
  "license": "Apache-2.0",
  "main": "lib/index.js",
  "name": "my-project",
  "pnpm": Object {},
  "scripts": Object {
    "build": "npx projen build",
    "bump": "npx projen bump",
    "clobber": "npx projen clobber",
    "compile": "npx projen compile",
    "default": "npx projen default",
    "eject": "npx projen eject",
    "eslint": "npx projen eslint",
    "package": "npx projen package",
    "post-compile": "npx projen post-compile",
    "post-upgrade": "npx projen post-upgrade",
    "pre-compile": "npx projen pre-compile",
    "projen": "npx projen",
    "release": "npx projen release",
    "test": "npx projen test",
    "test:watch": "npx projen test:watch",
    "unbump": "npx projen unbump",
    "upgrade": "npx projen upgrade",
    "watch": "npx projen watch",
  },
  "types": "lib/index.d.ts",
  "version": "0.0.0",
}
`;

exports[`sites/foo/.projen/deps.json 1`] = `
Object {
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
  "env": Object {
    "PATH": "$(pnpm -c exec \\"node --print process.env.PATH\\")",
  },
  "tasks": Object {
    "build": Object {
      "description": "Full release build",
      "name": "build",
      "steps": Array [
        Object {
          "spawn": "pre-compile",
        },
        Object {
          "spawn": "compile",
        },
        Object {
          "spawn": "post-compile",
        },
        Object {
          "spawn": "test",
        },
        Object {
          "spawn": "package",
        },
      ],
    },
    "compile": Object {
      "description": "Only compile",
      "name": "compile",
      "steps": Array [
        Object {
          "exec": "tsc --build",
        },
      ],
    },
    "default": Object {
      "description": "Synthesize project files",
      "name": "default",
    },
    "eslint": Object {
      "description": "Runs eslint against the codebase",
      "name": "eslint",
      "steps": Array [
        Object {
          "exec": "eslint --ext .ts,.tsx --fix --no-error-on-unmatched-pattern src test build-tools",
        },
      ],
    },
    "install": Object {
      "description": "Install project dependencies and update lockfile (non-frozen)",
      "name": "install",
      "steps": Array [
        Object {
          "exec": "pnpm i --no-frozen-lockfile",
        },
      ],
    },
    "install:ci": Object {
      "description": "Install project dependencies using frozen lockfile",
      "name": "install:ci",
      "steps": Array [
        Object {
          "exec": "pnpm i --frozen-lockfile",
        },
      ],
    },
    "package": Object {
      "description": "Creates the distribution package",
      "name": "package",
    },
    "post-compile": Object {
      "description": "Runs after successful compilation",
      "name": "post-compile",
    },
    "post-upgrade": Object {
      "description": "Runs after upgrading dependencies",
      "name": "post-upgrade",
    },
    "pre-compile": Object {
      "description": "Prepare the project for compilation",
      "name": "pre-compile",
    },
    "test": Object {
      "description": "Run tests",
      "name": "test",
      "steps": Array [
        Object {
          "exec": "jest --passWithNoTests --updateSnapshot",
          "receiveArgs": true,
        },
        Object {
          "spawn": "eslint",
        },
      ],
    },
    "test:watch": Object {
      "description": "Run jest in watch mode",
      "name": "test:watch",
      "steps": Array [
        Object {
          "exec": "jest --watch",
        },
      ],
    },
    "upgrade": Object {
      "description": "upgrade dependencies",
      "env": Object {
        "CI": "0",
      },
      "name": "upgrade",
      "steps": Array [
        Object {
          "exec": "pnpm update npm-check-updates",
        },
        Object {
          "exec": "npm-check-updates --upgrade --target=minor --peer --dep=dev,peer,prod,optional --filter=@types/jest,@types/node,@typescript-eslint/eslint-plugin,@typescript-eslint/parser,constructs,eslint-config-prettier,eslint-import-resolver-node,eslint-import-resolver-typescript,eslint-plugin-import,eslint-plugin-prettier,eslint,jest,jest-junit,npm-check-updates,prettier,projen,ts-jest,typescript",
        },
        Object {
          "exec": "pnpm i --no-frozen-lockfile",
        },
        Object {
          "exec": "pnpm update @types/jest @types/node @typescript-eslint/eslint-plugin @typescript-eslint/parser constructs eslint-config-prettier eslint-import-resolver-node eslint-import-resolver-typescript eslint-plugin-import eslint-plugin-prettier eslint jest jest-junit npm-check-updates prettier projen ts-jest typescript",
        },
        Object {
          "exec": "npx projen",
        },
        Object {
          "spawn": "post-upgrade",
        },
      ],
    },
    "watch": Object {
      "description": "Watch & compile in the background",
      "name": "watch",
      "steps": Array [
        Object {
          "exec": "tsc --build -w",
        },
      ],
    },
  },
}
`;

exports[`sites/foo/package.json 1`] = `
Object {
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
  "devDependencies": Object {
    "@types/jest": "*",
    "@types/node": "^16",
    "@typescript-eslint/eslint-plugin": "^6",
    "@typescript-eslint/parser": "^6",
    "constructs": "^10.0.0",
    "eslint": "^8",
    "eslint-config-prettier": "*",
    "eslint-import-resolver-node": "*",
    "eslint-import-resolver-typescript": "*",
    "eslint-plugin-import": "*",
    "eslint-plugin-prettier": "*",
    "jest": "*",
    "jest-junit": "^15",
    "npm-check-updates": "^16",
    "prettier": "*",
    "projen": "*",
    "ts-jest": "*",
    "typescript": "*",
  },
  "jest": Object {
    "clearMocks": true,
    "collectCoverage": true,
    "coverageDirectory": "coverage",
    "coveragePathIgnorePatterns": Array [
      "/node_modules/",
    ],
    "coverageReporters": Array [
      "json",
      "lcov",
      "clover",
      "cobertura",
      "text",
    ],
    "globals": Object {
      "ts-jest": Object {
        "tsconfig": "tsconfig.dev.json",
      },
    },
    "preset": "ts-jest",
    "reporters": Array [
      "default",
      Array [
        "jest-junit",
        Object {
          "outputDirectory": "test-reports",
        },
      ],
    ],
    "testMatch": Array [
      "<rootDir>/src/**/__tests__/**/*.ts?(x)",
      "<rootDir>/(test|src)/**/*(*.)@(spec|test).ts?(x)",
    ],
    "testPathIgnorePatterns": Array [
      "/node_modules/",
    ],
    "watchPathIgnorePatterns": Array [
      "/node_modules/",
    ],
  },
  "license": "UNLICENSED",
  "main": "lib/index.js",
  "name": "my-sub-project",
  "scripts": Object {
    "build": "npx projen build",
    "compile": "npx projen compile",
    "default": "npx projen default",
    "eslint": "npx projen eslint",
    "package": "npx projen package",
    "post-compile": "npx projen post-compile",
    "post-upgrade": "npx projen post-upgrade",
    "pre-compile": "npx projen pre-compile",
    "projen": "npx projen",
    "test": "npx projen test",
    "test:watch": "npx projen test:watch",
    "upgrade": "npx projen upgrade",
    "watch": "npx projen watch",
  },
  "types": "lib/index.d.ts",
  "version": "0.0.0",
}
`;
